bin/main.ml

```diff
@@ -28,27 +26,56 @@ let is_unique l =
   let () = List.iter (fun x -> Hashtbl.replace set x ()) l in
   len = Hashtbl.length set
 
+(* units for filter functions *)
+let has_same_size l s = List.length = s
+let rec has_duplicate_int_x l x = match l with
+  | [] -> true
+  | h :: t -> h = x && has_duplicate_int_x t x
 
 (* do the filter functions need to check for size? *)
 (* safety filter functions for duplicate *)
-let is_duplicate_prog1_safe l = is_duplicate l
-let is_duplicate_prog2_safe l = is_duplicate l
+let is_duplicate_prog1_safe l s x = 
+  has_same_size l s && has_duplicate_int_x l x
+let is_duplicate_prog2_safe l s x = 
+  has_same_size l s && has_duplicate_int_x l x
+let is_duplicate_prog2_safe l s x = 
```

Comment:
prog2 -> prog3... This makes me think that here, the duplication of code here may not be necessary